### PR TITLE
[inst-simplify] Update for OSSA

### PR DESCRIPTION
**Pull request by gottesmm**
I am splitting this out and fixing some issues at atrick's request.

### CODE REVIEW

File: lib/SIL/IR/SILInstruction.cpp [hunk 1]
```diff
@@ -1029,6 +1029,31 @@ SILInstruction::MemoryBehavior SILInstruction::getMemoryBehavior() const {
...
+    switch (li->getOwnershipQualifier()) {
+    case LoadOwnershipQualifier::Unqualified:
+    case LoadOwnershipQualifier::Trivial:
+    case LoadOwnershipQualifier::Take:
+      return MemoryBehavior::MayRead;
+    case LoadOwnershipQualifier::Copy:
+      return MemoryBehavior::MayHaveSideEffects;
+    }
+    llvm_unreachable("Covered switch isn't covered?!");
+  }
```
Comment (meg-gupta) [inline lib/SIL/IR/SILInstruction.cpp:1042]: We may also need to update  `MemoryBehaviorVisitor::visitLoadInst` and  `MemoryBehaviorVisitor::visitStoreInst` for [take] and [assign] ?

Comment (gottesmm) [reply lib/SIL/IR/SILInstruction.cpp:1042]: Yes. That is what I brought up here: https://github.com/apple/swift/pull/34543.

File: include/swift/SILOptimizer/Utils/ValueLifetime.h [hunk 2]
```diff
@@ -54,24 +55,47 @@ class ValueLifetimeAnalysis {
...
+
   /// Constructor for the value \p def with a specific range of users.
   ///
   /// We templatize over the RangeTy so that we can initialize
   /// ValueLifetimeAnalysis with misc iterators including transform
   /// iterators.
   template <typename RangeTy>
-  ValueLifetimeAnalysis(decltype(defValue) def, const RangeTy &userRange)
-      : defValue(def), userSet(userRange.begin(), userRange.end()) {
+  ValueLifetimeAnalysis(SILArgument *def, const RangeTy &useRange)
```
Comment (atrick) [inline include/swift/SILOptimizer/Utils/ValueLifetime.h:68]: Why is this renamed? `useRange` should have Operand types, but this still expects instruction pointers, or is there supposed to be a call to `getUser` here?

Also, why do we require a list of use Operands if we're only storing the set of user instructions?

File: lib/SIL/Utils/BasicBlockUtils.cpp [hunk 3]
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
...
+        continue;
+      }
+
+      visitedBlocks.insert(predBlock);
+      worklist.push_back(predBlock);
+    }
+  }
+}
+
+bool swift::completePostDominatingSet(
```
Comment (atrick) [inline lib/SIL/Utils/BasicBlockUtils.cpp:497]: Is there a reason this API doesn't return a move-only result
```
struct Result {
  bool hasLoop;
  SmallPtrSet<SILBasicBlock *, C>
};
```

File: lib/SIL/Utils/BasicBlockUtils.cpp [hunk 4]
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
...
+    SILBasicBlock *definingBlock, SILBasicBlock *reachableBlock,
```
Comment (atrick) [inline lib/SIL/Utils/BasicBlockUtils.cpp:498]: It seems odd for a utility to expect a single `reachableBlock`. In most cases don't we have a set of reachable blocks?

File: lib/SIL/Utils/BasicBlockUtils.cpp [hunk 5]
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
...
+    assert(definingBlock != reachableBlock);
+  }
+
+  void performDataflow();
+  void initializeDataflow();
+};
+
+} // anonymous namespace
+
+void PostDominanceCompletionSetComputer::initializeDataflow() {
```
Comment (atrick) [inline lib/SIL/Utils/BasicBlockUtils.cpp:409]: This algorithm does not have a separate "initialization" step, which implies something separate from the worklist processing. This function should just be a `pushPredecessors` utility called by the main `performDataflow`, once before entering the loop and once inside. It's actually alarming that it doesn't do the same thing, e.g. mark the predecessor visited.

File: lib/SIL/Utils/BasicBlockUtils.cpp [hunk 6]
```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
...
+      }
+      foundLoop = true;
+      continue;
+    }
+
+    worklist.push_back(pred);
+  }
+}
+
+void PostDominanceCompletionSetComputer::performDataflow() {
```
Comment (atrick) [inline lib/SIL/Utils/BasicBlockUtils.cpp:426]: I guess you could call the "data flow" but it's a little confusing because all it's doing is marking blocks visited. It just finds reachable blocks.

**Review by atrick — COMMENTED** [review]
Reviewed the first half, not including SimplifyInstruction or SILCombine

File: lib/SILOptimizer/Analysis/SimplifyInstruction.cpp [hunk 7]
```diff
@@ -727,45 +746,556 @@ SILValue swift::simplifyInstruction(SILInstruction *I) {
...
+    case ValueOwnershipKind::None:
+      continue;
+    case ValueOwnershipKind::Unowned:
+      llvm_unreachable("Unowned object can never be consumed?!");
+    }
+    llvm_unreachable("Covered switch isn't covered");
+  }
+}
+
+static ValueOwnershipKind findCompatibleNonNoneOwnershipKind(OperandOwnershipKindMap map) {
```
Comment (atrick) [inline lib/SILOptimizer/Analysis/SimplifyInstruction.cpp:989]: I'm worried to see new code dealing with `OwnershipKindMaps` instead of a simpler lifetime constraint model. The kind maps severely obscure the ownership model and I'm afraid it will be more difficult to eliminate the kind maps later if they keep getting propagated through APIs instead of being encapsulated behind a simple interface for "ownership constraints" ('owned', 'guaranteed', 'any') plus "lifetime constraints" ('consuming', 'immediate', 'scoped', 'propagating/escaping').

In this case, the entire `findCompatibleNonNoneOwnershipKind` helper should just be a trivial lattice 'join' over the new ownership kind and the operand's ownership constraint.

File: lib/SILOptimizer/Analysis/SimplifyInstruction.cpp [hunk 8]
```diff
@@ -727,45 +746,556 @@ SILValue swift::simplifyInstruction(SILInstruction *I) {
...
+      convertToUnowned.push_back(use);
+      continue;
+    }
+  }
+
+  // Ok, we have categorized our uses. First handle the conversion uses.
+  assert(onlyAcceptNone.empty() && convertToOwned.empty() && convertToGuaranteed.empty() && convertToUnowned.empty());
+
+  // Then handle the non-conversion uses.
+  switch (newKind) {
```
Comment (atrick) [inline lib/SILOptimizer/Analysis/SimplifyInstruction.cpp:1059]: I don't know why you need to separately switch over the operand ownership constraint above and switch over the new kind here. The new operand kind should be determined directly by joining the new value kind with the operand constraint. Then you switch over OwnershipKindPair reusing the same code in `::perform` but without any of the ".none" cases.

**Review by atrick — DISMISSED** [review]
I don't think you should handle .none as an ownership kind with special treatment. Converting to .none can be trivially handled before doing anything fancy. Converting from .none involves deciding the ownership of each use, which should be a trivial lattice join between the new kind and the operand's constraint. Then partitioning those uses uses into two (or three) separate kinds, and for each category reusing the same code "as-if" the original .none value were either owned or guaranteed.

File: lib/SIL/Utils/BasicBlockUtils.cpp (hunk 3 above)
Comment (gottesmm) [reply lib/SIL/Utils/BasicBlockUtils.cpp:497]: From our conversation: I'll add a comment explaining what is the input/output of the utility.

**Comment by gottesmm** [comment]
Going to rebase this.

**Comment by gottesmm** [comment]
I upstreamed and slimmed this down. I did it the way we talked about offline.

**Comment by gottesmm** [comment]
I think I also redid everything that were a problem.

File: include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h [hunk 9]
```diff
@@ -63,7 +63,8 @@ struct CanonicalizeInstruction {
   ///
   /// Warning: If the \p inst argument is killed and the client immediately
   /// erases \p inst, then it may be an invalid pointer upon return.
-  SILBasicBlock::iterator canonicalize(SILInstruction *inst);
+  SILBasicBlock::iterator canonicalize(SILInstruction *inst,
+                                       DeadEndBlocks &deadEndBlocks);
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h:67]: I think the issue was that I don't have access to the function in Canonicalization. So I need to pass it in.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
preset=buildbot,tools=RA,stdlib=RD,test=non_executable
@swift-ci Please test with preset macOS

**Comment by gottesmm** [comment]
I am going to have to fight a little bit with MSVC looks like.

**Comment by gottesmm** [comment]
I am really curious what the benchmark results turn out. I diffed the text and there isn't any difference at -O anyways.

**Comment by gottesmm** [comment]
If all of the tests pass, I am going to fix the windows issue and then smoke test this in.

File: include/swift/SILOptimizer/Utils/OwnershipUtils.h [hunk 10]
```diff
@@ -0,0 +1,51 @@
+//===--- OwnershipUtils.h -------------------------------------------------===//
```
Comment (atrick) [inline include/swift/SILOptimizer/Utils/OwnershipUtils.h:1]: Duplicate filename, easily distinguished as `OwnershipOptUtils.h`

File: lib/SILOptimizer/Utils/OwnershipUtils.cpp [hunk 11]
```diff
@@ -0,0 +1,804 @@
...
+#include "swift/SILOptimizer/Utils/ValueLifetime.h"
+
+using namespace swift;
+
+//===----------------------------------------------------------------------===//
+//                           Low Level RAUW Utility
+//===----------------------------------------------------------------------===//
+
+static SILBasicBlock::iterator
+replaceAllUsesAndEraseInner(SingleValueInstruction *svi, SILValue newValue,
```
Comment (atrick) [inline lib/SILOptimizer/Utils/OwnershipUtils.cpp:40]: Can there be just one copy of `replaceAllUsesAndEraseInner`?

File: lib/SILOptimizer/Utils/OwnershipUtils.cpp [hunk 12]
```diff
@@ -0,0 +1,804 @@
...
+  // replace any value with a ValueOwnershipKind::None value.
+  if (newOwnershipKind == OwnershipKind::None)
+    return true;
+
+  // If we are in Raw SIL, just bail at this point. We do not support
+  // ownership fixups.
+  if (oldValue->getModule().getStage() == SILStage::Raw)
+    return false;
+
+  // In constrast, if our old ownership kind is ValueOwnershipKind::None and
```
Comment (atrick) [inline lib/SILOptimizer/Utils/OwnershipUtils.cpp:786]: These comments are hard to parse because the description is the inverse of the condition being commented. You could either flip the condition in the code or delete the first sentence in this comment.

**Review by atrick — COMMENTED** [review]
Looks good overall. I'm still reviewing the RAUW utility and lifetime extender.

File: test/SILOptimizer/semantic-arc-opt-unchecked-ownership-conversion.sil [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+// Borrow scope is too small, we fail.
+//
+// CHECK-LABEL: sil [ossa] @owned_to_unowned_negative_1 : $@convention(thin) (@guaranteed Klass) -> () {
+// CHECK: unchecked_ownership_conversion
+// CHECK: } // end sil function 'owned_to_unowned_negative_1'
+sil [ossa] @owned_to_unowned_negative_1 : $@convention(thin) (@guaranteed Klass) -> () {
+bb0(%0 : @guaranteed $Klass):
+  %1 = copy_value %0 : $Klass
+  %2 = unchecked_ownership_conversion %1 : $Klass, @owned to @unowned
+  destroy_value %1 : $Klass
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply test/SILOptimizer/semantic-arc-opt-unchecked-ownership-conversion.sil:243]: Nothing. Unowned means that you have to retain it before you use it as an owned or guaranteed value. Otherwise, you can just pass it off.

File: include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (gottesmm) [reply include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h:67]: Ok. Let me look.

[the remark this answers is not in the document]
Comment (gottesmm) [reply include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h:67]: Fixed it locally.

File: include/swift/SILOptimizer/Utils/OwnershipUtils.h (hunk 10 above)
Comment (gottesmm) [reply include/swift/SILOptimizer/Utils/OwnershipUtils.h:1]: Good idea! Fixing!

File: include/swift/SILOptimizer/Utils/OwnershipUtils.h [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply include/swift/SILOptimizer/Utils/OwnershipUtils.h:13]: It is standard style according to LLVM style. I am just going with that.

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 15]
```diff
@@ -221,9 +221,17 @@ bool BorrowingOperand::visitLocalEndScopeUses(
...
+    auto *br = cast<BranchInst>(op->getUser());
+    for (auto *use : br->getArgForOperand(op)->getUses())
+      if (use->isLifetimeEnding())
+        if (!func(use))
+          return false;
     return true;
   }
+  }
+
+  llvm_unreachable("Covered switch isn't covered");
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/Utils/OwnershipUtils.cpp:234]: The reason to have this is so that if someone is lazy, ignores the compiler, we get an assert. I don't think it hurts and also shows that the code writer intended this code to always be covered. So I think it is reasonable.

File: lib/SILOptimizer/Mandatory/MandatoryCombine.cpp [hunk 16]
```diff
@@ -93,9 +95,9 @@ class MandatoryCombineCanonicalize final : CanonicalizeInstruction {
     changed = true;
   }
 
-  bool tryCanonicalize(SILInstruction *inst) {
+  bool tryCanonicalize(SILInstruction *inst, DeadEndBlocks &deadEndBlocks) {
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:98]: I fixed it.

File: lib/SILOptimizer/SemanticARC/OwnershipConversionElimination.cpp [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+using namespace semanticarc;
+
+//===----------------------------------------------------------------------===//
+//                            Top Level Entrypoint
+//===----------------------------------------------------------------------===//
+
+bool SemanticARCOptVisitor::visitUncheckedOwnershipConversionInst(
+    UncheckedOwnershipConversionInst *uoci) {
+  // Return false if we are supposed to only be running guaranteed opts.
+  if (ctx.onlyGuaranteedOpts)
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/SemanticARC/OwnershipConversionElimination.cpp:27]: Why don't we talk about this offline. I'll do a follow up patch to rename to whatever makes sense.

File: lib/SILOptimizer/Utils/OwnershipUtils.cpp [hunk 18]
```diff
@@ -0,0 +1,804 @@
+//===--- OwnershipUtils.cpp -----------------------------------------------===//
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Utils/OwnershipUtils.cpp:1]: Fixed.

File: lib/SILOptimizer/Utils/OwnershipUtils.cpp (hunk 12 above)
Comment (gottesmm) [reply lib/SILOptimizer/Utils/OwnershipUtils.cpp:786]: I fixed it.

File: lib/SILOptimizer/SemanticARC/OwnershipConversionElimination.cpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    return false;
+
+  // Then check if we are running tests and shouldn't perform this optimiuzation
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/SemanticARC/OwnershipConversionElimination.cpp:30]: Fixed.

File: lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp [hunk 20]
```diff
@@ -79,12 +79,8 @@ killInstAndIncidentalUses(SingleValueInstruction *inst,
...
 // intruction that wasn't erased.
 static Optional<SILBasicBlock::iterator>
-simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
-  // FIXME: temporarily bypass simplification untill all simplifications
-  // preserve ownership SIL.
-  if (inst->getFunction()->hasOwnership())
-    return None;
-
+simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass,
+                   DeadEndBlocks &deadEndBlocks) {
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:83]: Fixed.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
Done... sooo happy!
